### PR TITLE
[Game API] Add libretro Frame Time Callback

### PR DESCRIPTION
**Pull request by RobLoach**
Here's another attempt at moving over to Chrono. I've had issues compiling this on my Linux machine. I'll continue reading through the docs and see if I can come up with anything.

```
/* Notifies a libretro core of time spent since last invocation
 * of retro_run() in microseconds.
 *
 * It will be called right before retro_run() every frame.
 * The frontend can tamper with timing to support cases like
 * fast-forward, slow-motion and framestepping.
 *
 * In those scenarios the reference frame time value will be used. */
```

### CODE REVIEW

File: src/client.cpp [hunk 1]
```diff
@@ -358,6 +361,12 @@ GAME_ERROR RunFrame(void)
   if (!CLIENT)
     return GAME_ERROR_FAILED;
 
+  // Trigger the frame time callback before running the core.
+  uint64_t current = timer.microseconds();
+  int64_t delta = current - FRAME_TIME_LAST;
+  FRAME_TIME_LAST = current;
+  CLIENT_BRIDGE->FrameTime(delta);
```
Comment (garbear) [inline src/client.cpp]: delta will be a large value on the first call to RunFrame()

**Comment by garbear** [comment]
Code looks OK. What is needed to handle FF and slow motion?

File: src/client.cpp (hunk 1 above)
Comment (RobLoach) [reply src/client.cpp]: Mmm, we should likely attribute for that on initialization.

**Comment by RobLoach** [comment]
> What is needed to handle FF and slow motion?

After some investigation, I found that's handled elsewhere. This just simply returns the microsecond delta.

File: src/client.cpp (hunk 1 above)
Comment (RobLoach) [reply src/client.cpp]: Set it to timer.microseconds() over at https://github.com/kodi-game/game.libretro/pull/49/files#diff-34f48ad91ac6c202ac60b0348ae90e30R111

File: src/client.cpp [hunk 2]
```diff
@@ -109,6 +109,7 @@ ADDON_STATUS ADDON_Create(void* callbacks, void* props)
     }
 
     // Environment must be initialized before calling retro_init()
+    FRAME_TIME_LAST = timer.microseconds();
```
Comment (garbear) [inline src/client.cpp]: This is better, because now the first frame time will only be a second or two, but shouldn't it actually be zero?

Comment (RobLoach) [reply src/client.cpp]: Perhaps do an if statement to reset it? What would be the best solution do you think?

**Comment by garbear** [comment]
How about https://github.com/RobLoach/game.libretro/pull/1?

**Comment by RobLoach** [comment]
Nice work! That's a clean solution.

File: src/libretro/ClientBridge.cpp [hunk 3]
```diff
@@ -88,3 +89,13 @@ GAME_ERROR CClientBridge::AudioAvailable(void)
...
+
+GAME_ERROR CClientBridge::FrameTime(int64_t usec)
+{
+  if (!m_retro_frame_time_callback)
+    return GAME_ERROR_FAILED;
+
+  m_retro_frame_time_callback(usec);
+
+  return GAME_ERROR_NO_ERROR;
+}
```
Comment (garbear) [inline src/libretro/ClientBridge.cpp]: Can you fix the missing newline?

File: src/utils/Timer.cpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
+/*
+ *      Copyright (C) 2014-2016 Team Kodi
```
Comment (garbear) [inline src/utils/Timer.cpp]: 2018

File: src/utils/Timer.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
+/*
+ *      Copyright (C) 2016 Team Kodi
```
Comment (garbear) [inline src/utils/Timer.h]: 2018

File: src/utils/Timer.cpp [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ */
+
+#include "Timer.h"
+
+using namespace LIBRETRO;
+
+uint64_t Timer::microseconds()
+{
+  return std::chrono::duration_cast<std::chrono::microseconds>(m_clock.now().time_since_epoch()).count();
```
Comment (garbear) [inline src/utils/Timer.cpp]: Can you split this into two lines with a temporary const variable to make this line readable on small screens? The compiler will optimize out the temporary variable.

**Comment by RobLoach** [comment]
@garbear Any clue why it's failing on xcode?
https://travis-ci.org/kodi-game/game.libretro/jobs/473235450#L475-L487

```
Undefined symbols for architecture x86_64:
  "LIBRETRO::Timer::microseconds()", referenced from:
      _RunFrame in client.cpp.o
```

File: src/utils/Timer.cpp (hunk 4 above)
Comment (RobLoach) [reply src/utils/Timer.cpp]: Updated.

File: src/utils/Timer.h (hunk 5 above)
Comment (RobLoach) [reply src/utils/Timer.h]: Updated.

File: src/utils/Timer.cpp (hunk 6 above)
Comment (RobLoach) [reply src/utils/Timer.cpp]: Split into two lines! Thanks.

File: src/libretro/ClientBridge.cpp (hunk 3 above)
Comment (RobLoach) [reply src/libretro/ClientBridge.cpp]: Fixed.

**Comment by garbear** [comment]
You need to add the new files to CMakeLists.txt

**Comment by garbear** [comment]
LGTM. Can you squash?

**Comment by RobLoach** [comment]
Squashed.

**Comment by garbear** [comment]
A good follow-up to this would be exposing FF/RW through Kodi's Game API so we can adjust the frame time reported to the libretro core.

**Pull request merged**